### PR TITLE
ASM-2959 Moved code around and added job queue clear before export as well

### PR DESCRIPTION
**Pull request by asaenz**

### CODE REVIEW

File: lib/puppet/idrac/util.rb [hunk 1]
```diff
@@ -42,25 +43,118 @@ def self.wait_for_running_jobs
...
+          clear_job_queue(true)
+        end
+      end
+
+      def self.clear_job_queue(resetOnFailure=false)
+        Puppet.debug("Clearing Job Queue")
+        transport = get_transport
+        tries = 1
+        begin
+          endpoint={:host => transport[:host], :user => transport[:user], :password => transport[:password]}
```
Comment (gavin-scott) [inline lib/puppet/idrac/util.rb]: You could just use the transport as the endpoint

File: lib/puppet/idrac/util.rb [hunk 2]
```diff
@@ -42,25 +43,118 @@ def self.wait_for_running_jobs
         require 'asm/wsman'
         transport = get_transport
         Puppet.info("Checking for running jobs")
+        begin
+          10.times do
+            endpoint={:host=>transport[:host], :user=>transport[:user], :password=>transport[:password]}
```
Comment (gavin-scott) [inline lib/puppet/idrac/util.rb]: use transport as endpoint

File: lib/puppet/idrac/util.rb [hunk 3]
```diff
@@ -42,25 +43,118 @@ def self.wait_for_running_jobs
...
+        rescue Puppet::Idrac::JobClearError
+          Puppet.warning("Job queue still shows jobs exist after attempting to clear the job queue.")
+          reset if resetOnFailure
+        end
+      end
+
+      def self.wait_for_jobs_clear
+        transport = get_transport
+        Puppet.info("Waiting for job queue to be empty...")
+        endpoint={:host => transport[:host], :user => transport[:user], :password => transport[:password]}
```
Comment (gavin-scott) [inline lib/puppet/idrac/util.rb]: use transport as endpoint

File: lib/puppet/provider/importtemplatexml.rb [hunk 4]
```diff
@@ -274,7 +274,7 @@ def handle_missing_devices(xml_base, changes)
   #Helper function which just searches through the xml comments for BiosBootSeq value, since it will be commented out
   def find_bios_boot_seq(xml_base)
     uncommented = xml_base.at_xpath("//Attribute[@Name='BiosBootSeq']")
-    if !uncommented.nil?
+    unless uncommented.nil?
```
Comment (gavin-scott) [inline lib/puppet/provider/importtemplatexml.rb]: `if uncommented`?

File: lib/puppet/idrac/util.rb [hunk 5]
```diff
@@ -42,25 +43,118 @@ def self.wait_for_running_jobs
...
+              return
+            else
+              Puppet.debug("Job(s) still running: #{running_jobs}.")
+              Puppet.info("Waiting for currently running jobs to complete...")
+              sleep 30
+            end
+          end
+          raise("Timed out waiting for running jobs to complete.")
+        rescue
+          clear_job_queue(true)
```
Comment (gavin-scott) [inline lib/puppet/idrac/util.rb]: I would prefer to move the clear_job_queue piece out into a separate method that makes it clear that this is happening.

File: lib/puppet/idrac/util.rb [hunk 6]
```diff
@@ -42,25 +43,118 @@ def self.wait_for_running_jobs
...
+          end
+        rescue Puppet::Error => e
+          raise e if tries > 4
+          tries += 1
+          Puppet.info("Could not reset job queue.  Retrying in 30 seconds...")
+          sleep 30
+          retry
+        rescue Puppet::Idrac::JobClearError
+          Puppet.warning("Job queue still shows jobs exist after attempting to clear the job queue.")
+          reset if resetOnFailure
```
Comment (gavin-scott) [inline lib/puppet/idrac/util.rb]: Seems weird that after reset we haven't ensured that the job queue is empty.

File: lib/puppet/idrac/util.rb [hunk 7]
```diff
@@ -42,25 +43,118 @@ def self.wait_for_running_jobs
...
+              sleep 30
+            end
+          end
+          raise("Timed out waiting for running jobs to complete.")
+        rescue
+          clear_job_queue(true)
+        end
+      end
+
+      def self.clear_job_queue(resetOnFailure=false)
```
Comment (gavin-scott) [inline lib/puppet/idrac/util.rb]: A little comments on these methods as to what they are doing esp. w.r.t. error-handling behind the scenes would be great!

**Comment by asaenz** [comment]
@gavin-scott Made some changes to address your concerns.  Have another look.

**Pull request merged**